### ISSUE TITLE
Recursive handling of snippets

### ISSUE THREAD

**Issue by kshenoy** [SirVer/ultisnips#787]
Hi, I have the following snippets for SystemVerilog:

````
global !p
def complete(t, opts):
  if t:
    opts = [ m for m in opts if m.startswith(t) ]
  if len(opts) == 1:
    return opts[0][len(t):]
  elif len(opts) == 0:
    return ""
  return "  (" + " | ".join(opts) + ")"
endglobal


snippet al "always/always_comb/always_ff/always_latch"
always$1`!p snip.rv = complete( t[1], ['_comb', '_ff', '_latch'] )`
endsnippet

snippet always "always @( posedge clock ) .."
always @(${1:posedge clock}${2: or negedge reset}) begin
  ${0}
end
endsnippet
````

The first snippet is to choose between different versions of the always blocks such as always, always_ff, always_comb etc. and once I have selected it, I press Expand again to use the 2nd snippet.

Eg. I type `al`, press `<Tab>`. This types out `always` and gives me a completion menu with 3 options: _comb, _ff and _latch. Since, I want to expand the `always` form, I press `<Tab>` again and the completion options are replaced with the contents of the 2nd snippet. This is the right behavior. 

However, there's a quirk when I try to delete after pressing the 2nd `<Tab>`. In this case, I see the completion options from the 1st snippet again. It's like the 1st snippet is not able to recognize that it's job is done when the 2nd snippet is expanded.

Here's the full sequence (| indicates the position of the cursor):

    al|

Press <Tab>

    always|  (_comb | _ff | _latch)

I want to expand the always snippet so I press <Tab> again

    always @(posedge clock| or negedge reset) begin
    
    end

Now, I press <Esc> and delete these 3 lines. I expect to be left with nothing. However, the following appears:

      (_comb | _ff | _latch)

I'm using the default settings for expanding and moving to the next tabstop. Please let me know in case you need more info.

**Comment by seletskiy** [comment]
@kshenoy: Snippet engine works that way, so it tracks down all placeholders no matter how deep snippet nesting goes.

One workaround is to fool snippet engine and make it looks like that line inside snippet is replaced with something else, so engine will no longer track down placeholders in it:

```
pre_expand "snip.buffer[snip.line] = 'always'; snip.cursor.preserve()"
snippet always "always @( posedge clock ) .."
 @(${1:posedge clock}${2: or negedge reset}) begin
  ${0}
end
endsnippet
```

Bit ugly but should work.

**Comment by kshenoy** [comment]
Thanks! That worked; though I don't understand why or how. Can you elaborate on that?

**Comment by seletskiy** [comment]
@kshenoy: `pre_expand` will cause specified python code to be invoked right before snippet is expanded. `snip.buffer[snip.line] = 'always'` will effectively replace current line with `always`, then snippet engine will detect, that line from previous snippet was changed (e.g. replaced with static text), and snippet engine will remove all placeholders and interpolations from that line. `snip.cursor.preserve()` telling snippet engine to leave cursor where it is without any movement (it's required operation because limitations of vim API). Snipped body by itself doesn't have `always` string because we already placed `always` from `pre_expand` action.

In other words, code from `pre_expand` section kills previous snippet (dynamic text) by replacing it with static text.
